README.md
QtLongLongSpinBox
=================

The spin box (like QSpinBox) for qlonglong (64-bit signed integer) numbers:

![Screenshot](doc/examples/simple/img/screenshot001.png)

Example
=======
```c++
// main.cpp
#include <QApplication>
#include <QtLongLongSpinBox>

int main(int argc, char *argv[])
{
    QtLongLongSpinBox *widget = new QtLongLongSpinBox;
    QObject::connect(&app, SIGNAL(lastWindowClosed()), &app, SLOT(quit()));
    widget->resize(200, 20);
    widget->setValue(123456);
    widget->show();

    return app.exec();
}
```


CHANGELOG
ChangeLog
=========

v0.1.1
------
* behavior harmonized with QSpinBox

v0.1.0
------
* init version


src/qtlonglongspinbox.h
// Copyright 2014-2016, Durachenko Aleksey V. <durachenko.aleksey@gmail.com>
//
// This library is free software; you can redistribute it and/or
// modify it under the terms of the GNU Lesser General Public
// License as published by the Free Software Foundation; either
// version 2.1 of the License, or (at your option) any later version.
//
// This library is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public
// License along with this library; if not, write to the Free Software
// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
#ifndef QTLONGLONGSPINBOX_H
#define QTLONGLONGSPINBOX_H


#include <QAbstractSpinBox>
#include <QtGlobal>


class QtLongLongSpinBox : public QAbstractSpinBox
{
    Q_OBJECT
public:
    explicit QtLongLongSpinBox(QWidget *parent = 0);

    qlonglong value() const;

    QString prefix() const;
    void setPrefix(const QString &prefix);

    QString suffix() const;
    void setSuffix(const QString &suffix);

    QString cleanText() const;

    qlonglong singleStep() const;
    void setSingleStep(qlonglong val);

    qlonglong minimum() const;
    void setMinimum(qlonglong min);

    qlonglong maximum() const;
    void setMaximum(qlonglong max);

    void setRange(qlonglong min, qlonglong max);

public slots:
    void setValue(qlonglong value);

signals:
    void valueChanged(qlonglong i);
    void valueChanged(const QString &text);

protected:
    virtual void keyPressEvent(QKeyEvent *event);
    virtual void focusOutEvent(QFocusEvent *event);
    virtual void stepBy(int steps);
    virtual StepEnabled stepEnabled() const;
    virtual QValidator::State validate(QString & input, int &pos) const;

private:
    void lineEditEditingFinalize();
    void selectCleanText();

private:
    QString m_prefix;
    QString m_suffix;
    qlonglong m_singleStep;
    qlonglong m_minimum;
    qlonglong m_maximum;
    qlonglong m_value;

private:
    Q_DISABLE_COPY(QtLongLongSpinBox)
};


#endif // QTLONGLONGSPINBOX_H


src/qtlonglongspinbox.cpp
// Copyright 2014-2016, Durachenko Aleksey V. <durachenko.aleksey@gmail.com>
//
// This library is free software; you can redistribute it and/or
// modify it under the terms of the GNU Lesser General Public
// License as published by the Free Software Foundation; either
// version 2.1 of the License, or (at your option) any later version.
//
// This library is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public
// License along with this library; if not, write to the Free Software
// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
#include "qtlonglongspinbox.h"
#include <QLineEdit>
#include <limits>
#include <QEvent>
#include <QKeyEvent>


QtLongLongSpinBox::QtLongLongSpinBox(QWidget *parent) :
    QAbstractSpinBox(parent)
{
    m_minimum = std::numeric_limits<qlonglong>::min();
    m_maximum = std::numeric_limits<qlonglong>::max();
    m_value = 0;
    m_singleStep = 1;

    setValue(m_value);
}

qlonglong QtLongLongSpinBox::value() const
{
    return m_value;
}

void QtLongLongSpinBox::setValue(qlonglong expectedNewValue)
{
    const qlonglong newValue = qBound(m_minimum, expectedNewValue, m_maximum);
    const QString newValueString = QString::number(newValue);
    lineEdit()->setText(m_prefix + newValueString + m_suffix);
    if (m_value != newValue) {
        m_value = newValue;

        emit valueChanged(newValue);
        emit valueChanged(newValueString);
    }
}

QString QtLongLongSpinBox::prefix() const
{
    return m_prefix;
}

void QtLongLongSpinBox::setPrefix(const QString &prefix)
{
    m_prefix = prefix;

    setValue(m_value);
}

QString QtLongLongSpinBox::suffix() const
{
    return m_suffix;
}

void QtLongLongSpinBox::setSuffix(const QString &suffix)
{
    m_suffix = suffix;

    setValue(m_value);
}

QString QtLongLongSpinBox::cleanText() const
{
    return QString::number(m_value);
}

qlonglong QtLongLongSpinBox::singleStep() const
{
    return m_singleStep;
}

void QtLongLongSpinBox::setSingleStep(qlonglong step)
{
    m_singleStep = step;
}

qlonglong QtLongLongSpinBox::minimum() const
{
    return m_minimum;
}

void QtLongLongSpinBox::setMinimum(qlonglong min)
{
    m_minimum = min;
    if (m_maximum < m_minimum) {
        m_maximum = m_minimum;
    }

    setValue(m_value);
}

qlonglong QtLongLongSpinBox::maximum() const
{
    return m_maximum;
}

void QtLongLongSpinBox::setMaximum(qlonglong max)
{
    m_maximum = max;
    if (m_maximum < m_minimum) {
        m_maximum = m_minimum;
    }

    setValue(m_value);
}

void QtLongLongSpinBox::setRange(qlonglong min, qlonglong max)
{
    if (min < max) {
        m_minimum = min;
        m_maximum = max;
    }
    else {
        m_minimum = max;
        m_maximum = min;
    }

    setValue(m_value);
}

void QtLongLongSpinBox::keyPressEvent(QKeyEvent *event)
{
    switch (event->key()) {
    case Qt::Key_Enter:
    case Qt::Key_Return:
        selectCleanText();
        lineEditEditingFinalize();
    }

    QAbstractSpinBox::keyPressEvent(event);
}

void QtLongLongSpinBox::focusOutEvent(QFocusEvent *event)
{
    lineEditEditingFinalize();

    QAbstractSpinBox::focusOutEvent(event);
}

QAbstractSpinBox::StepEnabled QtLongLongSpinBox::stepEnabled() const
{
    if (isReadOnly()) {
        return 0;
    }

    StepEnabled se = 0;
    if (wrapping() || m_value < m_maximum) {
        se |= StepUpEnabled;
    }
    if (wrapping() || m_value > m_minimum) {
        se |= StepDownEnabled;
    }

    return se;
}

void QtLongLongSpinBox::stepBy(int steps)
{
    if (isReadOnly()) {
        return;
    }

    if (m_prefix + QString::number(m_value) + m_suffix != lineEdit()->text()) {
        lineEditEditingFinalize();
    }

    qlonglong newValue = m_value + (steps * m_singleStep);
    if (wrapping()) {
        // emulating the behavior of QSpinBox
        if (newValue > m_maximum) {
            if (m_value == m_maximum) {
                newValue = m_minimum;
            }
            else {
                newValue = m_maximum;
            }
        }
        else if (newValue < m_minimum) {
            if (m_value == m_minimum) {
                newValue = m_maximum;
            }
            else {
                newValue = m_minimum;
            }
        }
    }
    else {
        newValue = qBound(m_minimum, newValue, m_maximum);
    }

    setValue(newValue);
    selectCleanText();
}

QValidator::State QtLongLongSpinBox::validate(QString &input, int &pos) const
{
    // first, we try to interpret as a number without prefixes
    bool ok;
    const qlonglong value = input.toLongLong(&ok);
    if (input.isEmpty() || (ok && value <= m_maximum)) {
        input = m_prefix + input + m_suffix;
        pos += m_prefix.length();
        return QValidator::Acceptable;
    }

    // if string of text editor aren't simple number, try to interpret it
    // as a number with prefix and suffix
    bool valid = true;
    if (!m_prefix.isEmpty() && !input.startsWith(m_prefix)) {
        valid = false;
    }
    if (!m_suffix.isEmpty() && !input.endsWith(m_suffix)) {
        valid = false;
    }

    if (valid) {
        const int start = m_prefix.length();
        const int length = input.length() - start - m_suffix.length();

        bool ok;
        const QString number = input.mid(start, length);
        const qlonglong value = number.toLongLong(&ok);
        if (number.isEmpty() || (ok && value <= m_maximum)) {
            return QValidator::Acceptable;
        }
    }

    // otherwise not acceptable
    return QValidator::Invalid;
}

void QtLongLongSpinBox::lineEditEditingFinalize()
{
    const QString text = lineEdit()->text();

    // first, we try to read as a number without prefixes
    bool ok;
    qlonglong value = text.toLongLong(&ok);
    if (ok) {
        setValue(value);
        return;
    }

    // if string of text editor aren't simple number, try to interpret it
    // as a number with prefix and suffix
    bool valid = true;
    if (!m_prefix.isEmpty() && !text.startsWith(m_prefix)) {
        valid = false;
    }
    else if (!m_suffix.isEmpty() && !text.endsWith(m_suffix)) {
        valid = false;
    }

    if (valid) {
        const int start = m_prefix.length();
        const int length = text.length() - start - m_suffix.length();

        bool ok;
        const qlonglong value = text.mid(start, length).toLongLong(&ok);
        if (ok) {
            setValue(value);
            return;
        }
    }

    // otherwise set old value
    setValue(m_value);
}

void QtLongLongSpinBox::selectCleanText()
{
    lineEdit()->setSelection(m_prefix.length(),
                             lineEdit()->text().length()
                             - m_prefix.length()
                             - m_suffix.length());
}


examples/simple/main.cpp
// Copyright 2014-2016, Durachenko Aleksey V. <durachenko.aleksey@gmail.com>
//
// This library is free software; you can redistribute it and/or
// modify it under the terms of the GNU Lesser General Public
// License as published by the Free Software Foundation; either
// version 2.1 of the License, or (at your option) any later version.
//
// This library is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public
// License along with this library; if not, write to the Free Software
// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
#include <QApplication>
#include <QtLongLongSpinBox>


int main(int argc, char *argv[])
{
    QApplication app(argc, argv);

    QtLongLongSpinBox *widget = new QtLongLongSpinBox;   
    QObject::connect(&app, SIGNAL(lastWindowClosed()), &app, SLOT(quit()));
    widget->resize(200, 20);
    widget->setValue(123456);
    widget->show();

    return app.exec();
}
